packages/gapic-node-processing/src/library.ts

```diff
@@ -0,0 +1,271 @@
+  public isEsm: boolean;
+  public versions: string[] = [];
+  public clientsAndVersions: VersionsAndClients[] = [];
+  public srcPath: string;
+  public sourcePath: string;
+  public destinationPath: string;
+  public defaultVersion?: string = '';
+  public defaultVersionAndClients: VersionsAndClients = {
+    version: '',
+    clients: [],
+  };
+
+  /**
+   * @param isEsm Whether the library supports ES Modules.
+   * @param releaseLevel The current stability level of the library.
+   * @param defaultVersion The recommended version to use.
+   * @param sourcePath The path where the library is read from.
+   * @param destinationPath The path where the library is written to.
+   */
+  constructor(
+    sourcePath: string,
+    destinationPath: string,
+    defaultVersion?: string,
+    isEsm?: boolean,
+  ) {
+    this.isEsm = isEsm ?? false;
+    this.srcPath = isEsm ? ESM_SRC_PATH : SRC_PATH;
+    this.sourcePath = sourcePath;
+    this.destinationPath = destinationPath;
+    this.defaultVersion = defaultVersion;
+  }
+
+  /**
+   * Initializes async properties for the LibraryConfig instance.
+   * Call this after constructing the object.
+   */
+  public async initialize(): Promise<void> {
+    this.clientsAndVersions = await this.getClientsAndVersions();
+    this.versions = await this.getVersions();
+    if (!this.defaultVersion) {
+      this.defaultVersion = this.getHighestVersionWithPrecedence(this.versions);
+    }
+    this.defaultVersionAndClients = this.getDefaultVersionAndClients();
+  }
+
+  /**
+   * A getter to provide a list of all available versions.
+   */
+  public async getVersions(): Promise<string[]> {
+    return this.clientsAndVersions.map(x => x.version);
+  }
+
+  /**
+   * A getter to provide a list of clients and their corresponding versions.
+   */
+  public async getClientsAndVersions() {
+    const clientsAndVersions: VersionsAndClients[] = [];
+    const allVersionedLibraries = await getAllTopLevelDirectories(
+      this.sourcePath,
+    );
+    for (const directory of allVersionedLibraries) {
+      const versions = await getAllTopLevelDirectories(
+        path.join(this.sourcePath, directory, this.srcPath),
+      );
+      // even though this looks nested, it ends up being o(1) since
+      // we only have one directory per versioned library (the single
+      // version of the library)
+      for (const version of versions) {
+        const indexFile = path.join(
+          this.sourcePath,
+          directory,
+          this.srcPath,
+          version,
+          INDEX_PATH,
+        );
+        if (await fs.stat(indexFile)) {
+          const fileContent = await fs.readFile(indexFile, 'utf8');
+          const clientsRegexMatch = Array.from(
+            fileContent.matchAll(CLIENT_EXTRACTION_REGEX),
+          );
+          clientsAndVersions.push({
+            version,
+            clients: Array.from(
+              new Set(clientsRegexMatch.map((x: any) => x[1])),
```

Comment:
what does this line do?

Reply:
This line just makes sure we don't have any duplicates in the regex matching (with Set), and then creates an array from that set which is what the clients type is.

Reply:
made a suggestion to add this as a comment